### DISCUSSION TITLE
How to model a compound many-to-many relation

### DISCUSSION THREAD

**Discussion by 2color** [prisma/prisma#3022] [Q&A] [answered]
## Goal
I'm creating a data model to represent a grading system for courses. 

Key concepts
-  a `course` and `user` have a many-to-many relationship where a user can be associated with the role of a `STUDENT` or  `TEACHER` to a course.
- Each Course can have many Tests
- Each Test can have many TestReusults (for each student)
- Each TestResult is connected to a single Test and single user

## Questions

- Looking at the Client calls below, they all seem to work correctly. But just wanted to verify that I'm not breaking any conventions.

## Schema 
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  email   String          @unique
  id      Int             @default(autoincrement()) @id
  name    String
  courses UsersToCourse[]
}

model Course {
  id      Int             @default(autoincrement()) @id
  name    String
  members UsersToCourse[]
  tests   Test[]
}

model UsersToCourse {
  user      User     @relation(fields: [userId], references: [id])
  userId    Int
  course    Course   @relation(fields: [courseId], references: [id])
  courseId  Int
  createdAt DateTime @default(now())
  role      UserRole
  @@id([userId, courseId])
}

model Test {
  id         Int          @default(autoincrement()) @id
  updatedAt  DateTime     @updatedAt
  course     Course       @relation(fields: [courseId], references: [id])
  courseId   Int
  date       DateTime
  TestResult TestResult[]
}

model TestResult {
  id        Int      @default(autoincrement()) @id
  test      Test     @relation(fields: [testId], references: [id])
  testId    Int
  student   User     @relation(fields: [studentId], references: [id])
  studentId Int
  date      DateTime
  result Int // Percentage precise to one decimal point represented result * 10^-1
}

enum UserRole {
  STUDENT
  TEACHER
}

```

### Create a course 

```ts
  const course = await prisma.course.create({
    data: {
      name: 'Real World Prisma',
    },
  })
```

### Create a teacher and associate with the course

```ts
const teacher = await prisma.user.create({
    data: {
      email: 'daniel@prisma.io',
      name: 'Daniel',
      courses: {
        create: {
          course: { connect: { id: course.id } },
          role: 'TEACHER',
        },
      },
    },
  })
```

### Create a student and associate with the course

```ts
  const student1 = await prisma.user.create({
    data: {
      email: 'ada@prisma.io',
      name: 'Ada',
      courses: {
        create: {
          course: { connect: { id: course.id } },
          role: 'STUDENT',
        },
      },
    },
  })
```

### Create a course and associate students and teacher
```ts
const course2 = await prisma.course.create({
    data: {
      name: 'new course',
      members: {
        create: [
          {
            user: { connect: { email: 'ada@prisma.io' } },
            role: 'TEACHER',
          },
          {
            user: { connect: { email: 'daniel@prisma.io' } },
            role: 'STUDENT',
          },
          {
            user: { connect: { email: 'alvin@prisma.io' } },
            role: 'STUDENT',
          },
        ],
      },
    },
    include: {
      members: {
        include: { user: true },
      },
    },
  })
```

**Comment by mavilein** [comment, marked as answer]
First of some small feedback about the style of the schema:
* I personally find a schema hard to read if relation and scalar fields are intermingled. I usually only scan the relations at first sight to build a mental model. I personally always put them after the scalar fields.

And here is a bit of feedback about the general modeling. This is feedback that i would give someone who is indeed building an application in that domain. So maybe this is not relevant for you as you want to keep things simpler.
* The model `UsersToCourse` looks really good already as it contains additional useful information apart from the relation fields. Only the name could use some work imo. I would consider to use e.g. `CourseEnrollment`. Usually many to many relations are an important concept in a domain and the name should reflect that.
* On a semantic level I wonder how this system would deal with courses that are repeating. E.g. the current modeling prevents a user taking a course twice. Maybe there should be something like a `CourseRun` (couldn't find a better term). This model would sit in between `Course` and `UsersToCourse`. Also Tests should be scoped to a `CourseRun` so students can take a `Test` multiple times.

**Reply by 2color** [reply]
Thanks. That's immensely helpful.


@mavilein 
How do you think it's best to represent the test results (marks) that are calculated as a percentage? Intuitively, I'd use the [decimal numeric type](https://www.postgresql.org/docs/10/datatype-numeric.html) with PostgreSQL (precise to two decimal places). 

Looking at the [Prisma docs](https://www.prisma.io/docs/reference/database-connectors/postgresql#type-mapping-between-postgresql-to-prisma-schema), it's represented as a float in Prisma. I'm curious as to whether floating-point inaccuracies will be a problem.

**Reply by 2color** [reply]
Here's the most recent version of the schema, which addresses some of the points (have not implemented course runs). 

Additionally, I created two relations between `TestResult` to represent the `User` who graded and the `User` (student) who did the test 

```prisma
generator client {
  provider             = "prisma-client-js"
  experimentalFeatures = ["transactionApi", "aggregateApi"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  email       String             @unique
  id          Int                @default(autoincrement()) @id
  name        String
  social      Json?

  // Relation fields
  courses     CourseEnrollment[]
  testResults TestResult[]       @relation(name: "results")
  testsGraded TestResult[]       @relation(name: "graded")
}

model Course {
  id      Int                @default(autoincrement()) @id
  name    String

  // Relation fields
  members CourseEnrollment[]
  tests   Test[]
}

model CourseEnrollment {
  createdAt DateTime @default(now())
  role      UserRole

  // Relation Fields
  userId   Int
  courseId Int
  user     User   @relation(fields: [userId], references: [id])
  course   Course @relation(fields: [courseId], references: [id])

  @@id([userId, courseId])
}

model Test {
  id        Int      @default(autoincrement()) @id
  updatedAt DateTime @updatedAt
  name      String
  date      DateTime

  // Relation Fields
  courseId   Int
  course     Course       @relation(fields: [courseId], references: [id])
  testResult TestResult[]
}

model TestResult {
  id        Int      @default(autoincrement()) @id
  createdAt DateTime @default(now())
  result    Int // Percentage precise to one decimal point represented result * 10^-1

  // Relation Fields
  graderId  Int
  studentId Int
  testId    Int
  student   User @relation(name: "results", fields: [studentId], references: [id])
  gradedBy  User @relation(name: "graded", fields: [graderId], references: [id])
  test      Test @relation(fields: [testId], references: [id])
}

enum UserRole {
  STUDENT
  TEACHER
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
